File: Java/1-Arrays/Two_Sum.java
// t: O(N)
// s: O(N)
class Two_Sum {
    public int[] twoSum(int[] nums, int target) {
        
        if (nums.length == 0) return new int[]{};
        
        Map<Integer, Integer> table = new HashMap<>();
        
        for(int i = 0; i < nums.length; i++) {
            if (table.containsKey(target - nums[i])) {
                return new int[]{table.get(target - nums[i]), i};
            } 
            
            table.put(nums[i], i);
        }
        return new int[]{};
    }
}

File: Python/0_______ARRAY_______/Two_Sum.py
from ast import List
class Solution:
    # Algorithm
    # Create a hash_map, iterate over array with indexes. 
    # check complement = target - nums[i] is in hashtable, is so return [i, hashtable[complement]]
    # If not, record all the numbers in hash table with their indexes as values hashtable[nums[i]]=i
    #t: (N)
    #s: (N)
    def two_sum(self, nums: List[int], target: int) -> int:
        hash_map = {}
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in hash_map:
                return [i, hash_map[complement]]
            
            hash_map[nums[i]] = i

        
